### ISSUE TITLE
HTMLSlotElement assignedElements / assignedNodes not available in spec tests (reopen)

### ISSUE THREAD

**Issue by alvarocjunq** [ionic-team/stencil#2830]
Stencil version:

@stencil/core@2.3.0
I'm submitting a:

[x] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://stencil-worldwide.herokuapp.com/ or https://forum.ionicframework.com/

Current behavior:
In a .spec.tsx test, The mock HTML element returned from shadowRoot.querySelector('slot[name="items"]') does not have the assignedElements or assignedNodes functions on it.

Expected behavior:
It would have assignedElements and assignedNodes available, as per the HTMLSlotElement API.

I am able to call assignedElements() when running the component in the browsers I support.

Steps to reproduce:

menu-list.tsx

```ts
import { Component, ComponentInterface, Host, h } from '@stencil/core';

@Component({
  tag: 'menu-list',
  styleUrl: 'menu-list.css',
  shadow: true,
})
export class MenuList implements ComponentInterface {

  render() {
    return (
      <Host>
        <slot name="items"></slot>
      </Host>
    );
  }
}
```
menu-list.spec.tsx

```ts
describe('menu-list', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [MenuList],
      html:
      `<menu-list>
        <div slot="items">One</div>
        <div slot="items">Two</div>
        <div slot="items">Three</div>
      </menu-list>`,
    });

    const slotElement:HTMLSlotElement = page.root.shadowRoot.querySelector(
      'slot[name="items"]'
    );

    expect(slotElement.assignedElements()).toBeTruthy();
  });
});
```

I'm reopening the issue: [2606](https://github.com/ionic-team/stencil/issues/2606)
Cos is not solved until now and I have no access to reopen there =/

However, I have the same problem to test the assignedElements, cos I'm using this way into the component:

```ts
this.element.shadowRoot.querySelector('slot').assignedElements()
```

This line cause this error:
![image](https://user-images.githubusercontent.com/10416598/108484070-bd846580-729b-11eb-9042-096ca69dc034.png)

My dependencies:
```json
    "@stencil/core": "2.3.0",
    "@stencil/sass": "1.4.1",
    "@types/jest": "25.1.4",
    "jest": "25.1.0",
    "jest-cli": "25.1.0",
    "yargs": "15.3.0"
```

**Comment by wolfomat** [comment]
Wouldn't be it possible to mock HtmlSlotElement and spy on assignedElements?
As far as i see, the MockHTMLElement is either not complete, or it's just not intended to implement all methods of childs of HTMLElement?

As i am new to this topic i'm having no idea on how to perform this mock/spy construction

**Comment by runarberg** [comment]
Same thing happens to me in e2e tests

**Comment by alvarocjunq** [comment]
Any news from this issue?

**Comment by kirakishin** [comment]
same thing for me

**Comment by chrisweight** [comment]
This is a problem for me too, in both Specs and E2E tests

**Comment by elclanrs** [comment]
Same problem here. We use `assignedNodes()` in `componentDidLoad` so I can't even initialize the test as it errors out before I can write any expectations.

**Comment by driskull** [comment]
I'm seeing the same issue.

**Comment by driskull** [comment]
> I gave up and changed the implementation to not use assignedNodes() anymore =/

The problem with doing this is that it won't work in scenarios where one component wraps another one and the slotted element needs to pass from one to another.

For example, if you have `ComponentA` that uses `ComponentB` internally in its shadowRoot and you pass the a slotted element from `ComponentA` into a slot of `ComponentB`.

You won't be able to use `querySelector` because the slotted elements are not in the same rootNode.

```
// ComponentA: uses ComponentB internally
render() {
    return (
        <ComponentB>
          <slot />
       </ComponentB>
    );
  }

// ComponentB: needs to get assigned elements of the slot. querySelector won't work in here if ComponentA is used since the slotted elements are in light DOM and not in the shadowRoot. It would be much simpler to just use assignedElements to get the elements here.
render() {
    return (
        <slot />
    );
 }
```

**Comment by sftreinman** [comment]
Re: the example from the post above, I had a similar scenario. 

Child component: `this.template.querySelector('slot[name="footer"]').assignedElements({flatten: true});`
This retrieved the elements from the slot for me... basically the same thing result alvarocjunq ended up using.

**Comment by driskull** [comment]
Yep, querying for the slot and getting its assignedElements works perfectly. The only issue is that it won't work in e2e tests so we can't test the functionality properly.

**Comment by leonheess** [comment]
Solved this by changing `.assignedNodes()` to `.assignedNodes?.()` in the code.

**Comment by iliocatallo** [comment]
I'm experiencing the same problem in stencil 3.0.0: `assignedNodes()` and `assignedElements()` are both `undefined` on slot elements.

**Comment by renet** [comment]
A workaround I found that helps preventing `assignedElements is not a function` and similar errors, even when they are used in early lifecycle methods, is the following:

```js
beforeAll(() => {
  Object.defineProperty(MockHTMLElement.prototype, 'assignedElements', {
    value: () => [], // `jest.fn()` works as well, but will result in just one shared function for all elements
    configurable: true,
    writable: true,
  });
  Object.defineProperty(MockHTMLElement.prototype, 'assignedNodes', {
    value: () => [],
    configurable: true,
    writable: true,
  });
});
```

for usage in later lifecycle methods, when you need to define the return value of those methods for proper testing, you can do as follows:

```js
const slotElement = page.root.shadowRoot.querySelector<HTMLSlotElement>('slot');
const childElements = Array.from(page.root.children);  // or similar

slotElement.assignedElements = jest.fn().mockReturnValue(childElements);
```

Hope this helps. If anyone has an idea how to make the mocked function return the actually assigned elements/nodes and how to mock those methods for all tests at once (instead of only for the current test file), please let me know. Didn't invest any further.

**Comment by OscarVagleStatnett** [comment]
One way to test if slotted tags are reflected into the DOM could be to listen for the "slotchange" event like this:

`

    let slot= page.root.shadowRoot.querySelector('...slot[name="tags"]')
    // Listen to slotchange event
    let slotChangeFired = false;
    firstSlot.addEventListener('slotchange', () => {
      slotChangeFired = true;
    });

    // Add tags with slot attributes
    const tagsSlot = document.createElement('div');
    tagsSlot.setAttribute('slot', 'tags');
    page.root.appendChild(tagsSlot);

    // Trigger slotchange event
    firstSlot.dispatchEvent(new Event('slotchange'));

    // Assert slotchange event fired
    expect(slotChangeFired).toBe(true);`
    
    
This tests if slots are added into the DOM, but doesn't test if slotted tags are inserted into the correct place in the DOM. However, this is something that can be used.
